### PR TITLE
Synchronous AMQP Close

### PR DESCRIPTION
**Pull request by LarryOsterman**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/eventhubs/azure-messaging-eventhubs/src/consumer_client.cpp
Comment (antkmsft) [inline sdk/eventhubs/azure-messaging-eventhubs/src/consumer_client.cpp:58]: If I understand correctly, this parameter has a default value, so could/should be written as `Close()`?

File: sdk/eventhubs/azure-messaging-eventhubs/src/partition_client.cpp
Comment (antkmsft) [inline sdk/eventhubs/azure-messaging-eventhubs/src/partition_client.cpp:178]: Could/should be written as `Close()`?

File: sdk/core/azure-core-amqp/samples/internal/eventhub_get_properties_sample/eventhub_get_eventhub_properties_sample.cpp
Comment (antkmsft) [inline sdk/core/azure-core-amqp/samples/internal/eventhub_get_properties_sample/eventhub_get_eventhub_properties_sample.cpp:77]: Could/should be written as `Close()`?

File: sdk/core/azure-core-amqp/CHANGELOG.md
Comment (antkmsft) [inline sdk/core/azure-core-amqp/CHANGELOG.md:11]: This change is in `_internal` surface, so I think you don't have to add this changelog line.

File: sdk/core/azure-core-amqp/src/amqp/management.cpp
Comment (antkmsft) [inline sdk/core/azure-core-amqp/src/amqp/management.cpp:136]: Is this exception- and thread-safe?
I think I at least saw some throw expressions in these Close methods' implementations. Maybe it is ok.

**Review by antkmsft — APPROVED** [review]

File: sdk/core/azure-core-amqp/src/amqp/management.cpp
Comment (antkmsft) [inline sdk/core/azure-core-amqp/src/amqp/management.cpp:137]: Is this exception- and thread-safe?
I think I at least saw some throw expressions in these Close methods' implementations. Maybe it is ok.

File: sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md
Comment (antkmsft) [inline sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md:10]: Purely from reading these lines, I don't see how adding an optional parameter is a breaking change.

File: sdk/core/azure-core-amqp/CHANGELOG.md
Comment (antkmsft) [inline sdk/core/azure-core-amqp/CHANGELOG.md:11]: This one as well, you even say it should not affect the existing code, yet you list it in Breaking Changes.

File: sdk/eventhubs/azure-messaging-eventhubs/src/consumer_client.cpp
Comment (LarryOsterman) [reply sdk/eventhubs/azure-messaging-eventhubs/src/consumer_client.cpp:58]: It could. This change was made before it was defaulted. I didn't bother to change it because ultimately it doesn't matter.

File: sdk/eventhubs/azure-messaging-eventhubs/src/partition_client.cpp
Comment (LarryOsterman) [reply sdk/eventhubs/azure-messaging-eventhubs/src/partition_client.cpp:178]: Could, not sure it matters though.

File: sdk/core/azure-core-amqp/samples/internal/eventhub_get_properties_sample/eventhub_get_eventhub_properties_sample.cpp
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/samples/internal/eventhub_get_properties_sample/eventhub_get_eventhub_properties_sample.cpp:77]: Could, same answer as the previous 3 answers.

File: sdk/core/azure-core-amqp/CHANGELOG.md
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/CHANGELOG.md:11]: Azure Core has changes to the _internal surface in the changelog, why is AMQP different?

File: sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md
Comment (LarryOsterman) [reply sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md:10]: Adding the optional parameter isn't necessarily a breaking change. Changing the semantics of `Close` to block waiting on the response to the `DETACH` AMQP message may be.

I would rather be more conservative in breaking change notifications than assume that a behavior change won't have any side effects.

File: sdk/core/azure-core-amqp/CHANGELOG.md
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/CHANGELOG.md:11]: The change from non-blocking close to blocking close can theoretically break clients, I'd rather call out changes as breaking than not, especially at this point in development.

Comment (antkmsft) [reply sdk/core/azure-core-amqp/CHANGELOG.md:11]: Does it? The same comment applies there then.

Comment (antkmsft) [reply sdk/core/azure-core-amqp/CHANGELOG.md:11]: Finally, up to you as an author. All I said, "you don't have to".

File: sdk/eventhubs/azure-messaging-eventhubs/src/consumer_client.cpp
Comment (antkmsft) [reply sdk/eventhubs/azure-messaging-eventhubs/src/consumer_client.cpp]: I'd recommend to put it simply as `()` rather than `({})` - less questioning, if it matters and why it is written this way, which could take a few minutes to confirm (when you are investigating strange bugs, and don't know where to start, you question everything).

I would think, that there are two overloads of `Close()` - one with a parameter, another one is parameterless. And with syntax your intent is to call the one with a parameter, rather than the parameterless one.

**Comment by antkmsft** [comment]
Can this fix #5233?

File: sdk/core/azure-core-amqp/src/amqp/message_receiver.cpp
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/src/amqp/message_receiver.cpp:547]: A null response from WaitForResult implies that the context was cancelled, or something else happened which is effectively similar to cancelled. But the context might not actually have been cancelled (for instance, if the link was disconnected from the sender, we treat that as a cancelled operation).

**Comment by LarryOsterman** [comment]
> Can this fix #5233?

Possibly, I honestly don't know.

File: sdk/core/azure-core-amqp/CHANGELOG.md
Comment (LarryOsterman) [reply sdk/core/azure-core-amqp/CHANGELOG.md:11]: For AMQP, the entire functional API surface is in _internal, so I think it belongs.

And yes, for instance we have documented changes to the logging infrastructure, which is all internal (changing the format of log strings, updating log levels for individual messages, etc).

**Review by antkmsft — APPROVED** [review]
I looked at your recent commits made after my previous approval, all still looks good.

**Pull request merged**
